src/types.jl

```diff
@@ -414,15 +421,22 @@ AGray32(g::AbstractGray, alpha = 1) = AGray32(gray(g), alpha)
 # Note: with the exceptions of `alphacolor` and `coloralpha`, all
 # traits in the rest of this file are intended just for internal use
 
+# The following should be in traits.jl but we need it now.
+# Return the number of components in the color
+# Note this is different from div(sizeof(c), sizeof(eltype(c))) (e.g., XRGB)
+length(c::Colorant) = length(typeof(c))
+length(::Type{C}) where C<:(Colorant{T,N} where T) where N = N
+
 const color3types = map(s->getfield(ColorTypes,s),
-  filter(names(ColorTypes, all=false)) do s
-    isdefined(ColorTypes, s) || return false
-    t = getfield(ColorTypes, s)
-    isa(t, Type) && t <: Colorant && !isabstracttype(t) && length(fieldnames(t))>1
-  end)
+                        filter(names(ColorTypes, all=false)) do s
+                            isdefined(ColorTypes, s) || return false
+                            t = getfield(ColorTypes, s)
+                            isa(t, Type) && t <: Color && !isabstracttype(t) && length(t) == 3
+                        end
+                        )
```

Comment:
I think the following is simpler and it is better to use `unique`.
```julia
filter(map(s->getfield(ColorTypes, s), names(ColorTypes, all=false))) do s
    isa(s, Type{<:Color3}) && !isabstracttype(s)
end |> unique
```

Reply:
I took some of this, good suggestion.

But:
```julia
julia> module MissingDef
       export foo
       end
Main.MissingDef

julia> names(MissingDef)
2-element Array{Symbol,1}:
 :MissingDef
 :foo       

julia> getfield(MissingDef, :foo)
ERROR: UndefVarError: foo not defined
Stacktrace:
 [1] top-level scope at REPL[25]:1
```

Reply:
In general, I think it's good that errors occur. We can notice the problems and improve it. :smile:
Is there a reason that the definition should be missing in ColorTypes.jl?

Reply:
I guess I always think of these as a pair, probably because I often write code on the compiler side of Julia where it's quite important to avoid making assumptions that all packages are perfect. But sure, as long as there's a suitable comment we could afford to be fragile here.

Reply:
Oh, actually, we need my version after all: any exported name in this package that hasn't yet been loaded will cause an error:

```julia
julia> include("runtests.jl")
[ Info: Precompiling ColorTypes [3da002f7-5984-5a60-b8a6-cbb66c0b333f]
ERROR: LoadError: LoadError: UndefVarError: alpha not defined
Stacktrace:
 [1] #2 at /home/tim/.julia/dev/ColorTypes/src/types.jl:430 [inlined]
 [2] iterate at ./generator.jl:47 [inlined]
...
```
`alpha` isn't defined until `trails.jl` gets loaded.

Reply:
I have the opposite conclusion. This means that it depends on the order of loading, so your version has the "potential" risk of the absence of types. As you say, we have to think that as a pair.

Reply:
I guess it's not as big a deal either way now. Older versions of this package defined some of the traits in terms of `color3types`, so you had to define its value early (before some traits were defined).

So it would be fine to change this if you want to move it to the end of the module definition.